content/client-lib-development-guide/features.textile

```diff
@@ -599,6 +599,7 @@ h3(#realtime-channel). Channel
 ** @(RTL7d)@ Messages delivered are automatically decoded based on the @encoding@ attribute; see REST @Channel@ encoding features. Tests should exist to publish and subscribe to encoded messages using the "AES 128":https://github.com/ably/ably-common/blob/main/test-resources/crypto-data-128.json and "AES 256":https://github.com/ably/ably-common/blob/main/test-resources/crypto-data-256.json fixture test data
 ** @(RTL7e)@ If a message cannot be decoded or decrypted successfully, it should be delivered to the listener with the @encoding@ attribute set indicating the residual encoding state, and an error should be logged
 ** @(RTL7f)@ A test should exist ensuring published messages are not echoed back to the subscriber when @echoMessages@ is set to false in the @Realtime@ library constructor
+** @(RTL7g)@ Subscribe with a function argument subscribes a listener which only fires when the function returns @true@
```

Comment:
I do think we may have talked our way out of this being particularly convenient.

My suggestion in [Proposal: Message References (May 2022)](https://docs.google.com/document/d/1gptZ5meelzeWhXLB5EvDsUMRb1Af7EYR1XPYRUbKO-0/edit#heading=h.i8z9qmamfguq) was to support specific functions for things like "subscribe to reactions to this specific message":

```
channel
  .subscriber()
  .refId("v1b25XrTDg:0")
  .refType("com.ably.reaction.v1")
  .forEach(msg => { … })
```

whereas with this function as described here, that'd be something like:

```
channel.subscribe(
  msg => msg.ref && msg.ref.id === "v1b25XrTDg:0" && msg.ref.type === "com.ably.reaction.v1",
  msg => { ... },
)
```

which makes me wonder if that's really much more convenient than:

```
channel.subscribe(msg => {
  if(!msg.ref || msg.ref.id !== "v1b25XrTDg:0" || msg.ref.type !== "com.ably.reaction.v1") {
    return;
  )
  ...
)
```

Or, do we imagine providing convenience functions which return filters? If so, could they be composed?

Reply:
My main issue with the builder style is that it's completely different from how any other part of the SDKs work (as far as I'm aware) and would feel out of place to me, unless we were looking to move subscriptions over to this style in the future and deprecate `subscribe()`. 

Perhaps a builder for filter functions similar to your original concept, except it returns a `(Message)=>bool` that can be inserted as a filter would be the best of both worlds? e.g `channel.subscribe(filter.withRef("v1b25XrTDg:0").withType("com.ably.reaction.v1"), (msg)=>{})`  
  
Aside from that, the lack of usefulness of a filter function makes me wonder what benefit we actually get from modifying the SDK here at all. Perhaps all the logic should be within a higher level ably-chat SDK?